compiler/src/dotty/tools/dotc/transform/TypeTestsCasts.scala

```diff
@@ -122,9 +122,19 @@ object TypeTestsCasts {
           case _                   => recur(defn.AnyType, tpT)
         }
       case tpe: AppliedType     =>
-        // first try withou striping type parameters for performance
-        isClassDetermined(X, tpe)(ctx.fresh.setNewTyperState()) ||
-        isClassDetermined(stripTypeParam(X), tpe)(ctx.fresh.setNewTyperState())
+        X.widen match {
+          case OrType(tp1, tp2) =>
+            // This case is required to retrofit type inference,
+            // which cut constraints in the following two cases:
```

Comment:
I'm familiar with `TypeComparer#either` but I'm not sure I get what you are saying...
```
This case is required to match type inference, which simplifies constraints in the following two cases:
```

But if that's the reason, what about `T1 & T2 <:< T3`? Do we need to handle that somehow?

Reply:
`T1 & T2 <:< T3` is handled in the top-level match on `P`